### DISCUSSION TITLE
Save GraphQL queries as REST endpoint

### DISCUSSION THREAD

**Discussion by denisyilmaz** [craftcms/cms#10908] [Ideas] [answered]
For several projects we are using Craft CMS together with [Hasura](https://hasura.io) to connect both GraphQL Schemas into one single API. Which is working amazingly well.

In a recent version, Hasura introduced a feature to save GraphQL queries to a custom REST Api endpoint so common queries can be easily be fetched and cached without the need to have big `gql` queries saved on the client, which can blow up a website pretty fast. 

I love this feature as it allows a nice refactoring of projects as not all projects need the flexibility GraphQL has to offer and would benefit of actually using simple REST Api endpoints to fetch different data. Hasura also managed to allow mutations via POST requests as well.

Is this something that Craft CMS might be interested in implementing?

Here are some screenshots from Hasura:

## GraphiQL example from within Hasura
Example shows querying entries from Craft CMS directly from Hasura. Craft CMS is attached as a "remote Schema".
You can see the "REST" button in the nav bar:

<img width="875" alt="image" src="https://user-images.githubusercontent.com/4189607/163144151-c0cd2483-be2a-412f-aa1a-39b53f967ea1.png">

## Saving Query as a REST endpoint

![image](https://user-images.githubusercontent.com/4189607/163144653-0bcc9f37-08da-4dda-ac47-aeecc17c5f5a.png)
![image](https://user-images.githubusercontent.com/4189607/163145655-d1aab138-f81b-490d-ba0c-b6bae21e1c76.png)

**Comment by andris-sevcenko** [comment, marked as answer]
This is covered by #8964, it's better to follow along in there.

**Reply by denisyilmaz** [reply]
isnt the idea i'm outlining here (and which hasura implemented) something different compared to persisted queries? as far as i understood persisted queries the client still needs to have the original query at its disposal in case the backend does not know the query hash, right?
My idea here would be to have a button inside the Craft CMS CP that allows to save a query to a REST Api endpoint.

**Reply by andris-sevcenko** [reply]
I might have missed something in the discussion I linked, but here's how I see how it would play out:

* In the GraphiQL IDE, add a new button "Persist query" or something like that, which would take the current query, store it server-side and generate a hash for it.
* You'd be able to list the persisted queries in GraphiQL IDE as well.
* When querying, you just pass in the hash as part of the header, Craft looks up the query based on the hash and executes it (token permitting)

The whole goal of persisted queries is precisely to avoid having queries in the frontend - this allows for smaller frontend assets as well as, if combined with a strict approach of "only persisted queries", adds another layer of security where the API consumers can't snoop around that much.

Does that make sense and link the two discussions together?

**Reply by jasonmccallister** [reply]
@denisyilmaz I might want to set some time aside with you to discuss this in the near future and more specifically how you are using Hasura with Craft (if you are up for it).

**Reply by denisyilmaz** [reply]
Hi @jasonmccallister, absolutely. Just hit me up via email! The next few weeks are a bit overload, but then I would be up to discuss this topic!
